woocommerce/woocommerce-ios
# Filter Products by Category: synchronize remotely the Filter Product Category

Part of https://github.com/woocommerce/woocommerce-ios/issues/5159

## Description
In the context of https://github.com/woocommerce/woocommerce-ios/issues/5159, the user can select a product category to filter the products, what causes the category being stored locally so the filter is still enabled when the user restarts the application. With this PR I want to handle the case when that product category is deleted externally (e.g from wp-admin), an event that would trigger an inconsistent state when going back to the products list or opening the Filter view.
To address that, we compare the locally stored product category filter with is remote counterpart, removing the filter if the category was deleted.

## Decisions
In the the implementation suggested with this PR, we do not wait for the ProductCategory to be synchronized to apply the filters, but we show them as soon as they retrieved locally. After that we synchronize the `ProductCategory` asynchronously, and then update the view (filters + products) **only** if the `ProductCategory` changed or was removed. This way we can speed up the products load with the filters process.

## Notes
This logic is added to the `ProductsViewController` as I did not find any view model in charge of implementing the presentation logic of the view, thus assuming that MVC is the architecture here.

## Changes
- Add the relevant methods in `ProductCategoryStore` and `ProductCategoriesRemote` to retrieve the remote information of a single product category given its `categoryID`
- Create the  cases for `DotComError` and `ProductCategoryActionError` to encapsulate the event of a failure because the resource was not there
- Once the local product settings are retrieved when applying the filters in the `ProductsViewController`, the locally `ProductCategory` is synchronized against its remote entity. If the the category was deleted, we update the filters product information, and remove the filter locally.
- Unit tests are added. New naming convention and asynchronous testing approach are applied to all the tests in the modified files.

## Tests
- Open app
- Go to products
- Open filter 
- Select Product Category
- Select a Product Category
-  Press "Show products" so the Product Category filter is stored
- You can see that the filter number on top of the products list is updated to include the Product Category filter
- Close the app
- Go to your website wp-admin
- Delete that category
- Open the app again and go to products
- After the filter is synchronized, the Filters number does not include the Product Category filter
- Open filter. There is no Product Category selected, _Any_ is selected

Update release notes:
